### PR TITLE
[TACACS+] Add TACACS per-command accounting test case.

### PR DESCRIPTION
**Pull request by liuh-80**
### Description of PR
    Add new E2E test cases for TACACS per-command accounting.
        These UT depends on the per-command accounting feature in sonic-buildimage repo and AAA accounting config command in sonic-utility repo.

### Type of change
- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

#### What is the motivation for this PR?
    Add E2E test case coverage for the per-command accounting feature.

#### How did you do it?
   Add new E2E test case to cover user scenarios.

#### How did you verify/test it?
   Run new UT make sure they are all pass.
   Make sure all current UT not break during merge validation.

#### Any platform specific information?
   N/A

### Documentation 
https://github.com/Azure/SONiC/blob/master/doc/aaa/TACACS%2B%20Design.md

### CODE REVIEW

File: tests/tacacs/test_accounting.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+@pytest.fixture(autouse=True, scope="module")
+def check_image_version(duthost):
+    """Skips this test if the SONiC image installed on DUT is older than 202112
+    Args:
+        duthost: Hostname of DUT.
+    Returns:
+        None.
+    """
+    skip_version(duthost, ["201811", "201911", "202012", "202106"])
```
Comment (qiluo-msft) [inline tests/tacacs/test_accounting.py:81]: >skip_version [](http://example.com/codeflow?start=4&length=12)

I notice more discussion #4736

File: tests/tacacs/test_accounting.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+from .test_authorization import ssh_connect_remote, ssh_run_command
+from .utils import stop_tacacs_server, start_tacacs_server
+from tests.common.errors import RunAnsibleModuleFail
+from tests.common.helpers.assertions import pytest_assert
+from tests.common.utilities import skip_version
+
+pytestmark = [
+    pytest.mark.disable_loganalyzer,
+    pytest.mark.topology('any'),
+    pytest.mark.device_type('vs')
```
Comment (qiluo-msft) [inline tests/tacacs/test_accounting.py:14]: >vs [](http://example.com/codeflow?start=29&length=2)

Will it run on true DUT?

File: tests/tacacs/test_accounting.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+]
+
+logger = logging.getLogger(__name__)
+
+def cleanup_tacacs_log(ptfhost, rw_user_client):
+    try:
+        ptfhost.command('rm /var/log/tac_plus.acct')
+    except RunAnsibleModuleFail as e:
```
Comment (qiluo-msft) [inline tests/tacacs/test_accounting.py:22]: >e [](http://example.com/codeflow?start=35&length=1)

Do you want to log something about `e`?

File: tests/tacacs/test_accounting.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        logger.info("/var/log/tac_plus.acct does not exist.")
+
+    res = ptfhost.command('touch /var/log/tac_plus.acct')
+    logger.info(res["stdout_lines"])
+
+    ssh_run_command(rw_user_client, 'sudo truncate -s 0 /var/log/syslog')
+
+def check_tacacs_server_log_exist(ptfhost, duthost, creds_all_duts, command):
+    username = creds_all_duts[duthost]['tacacs_rw_user']
+    sed_command = "sed -nE '/	{0}	.*	cmd=.*{1}/P' /var/log/tac_plus.acct".format(username, command)
```
Comment (qiluo-msft) [inline tests/tacacs/test_accounting.py:32]: >sed [](http://example.com/codeflow?start=19&length=3)

Could you add code comment to explain each `sed`?

File: tests/tacacs/test_authorization.py [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+        if exp_val in l:
+            content_exist = True
+            break
+    pytest_assert(content_exist)
+
+@pytest.fixture
+def remote_user_client(duthosts, enum_rand_one_per_hwsku_hostname, creds_all_duts):
+    duthost = duthosts[enum_rand_one_per_hwsku_hostname]
+    dutip = duthost.host.options['inventory_manager'].get_host(duthost.hostname).vars['ansible_host']
+    ssh_client = ssh_connect_remote(dutip, creds_all_duts[duthost]['tacacs_authorization_user'],
```
Comment (qiluo-msft) [inline tests/tacacs/test_authorization.py:54]: >ssh_client [](http://example.com/codeflow?start=4&length=10)

Use "with statement" ?
ref: [https://stackoverflow.com/a/42299058/2514803](https://stackoverflow.com/a/42299058/2514803)

File: tests/tacacs/tac_plus.conf.j2 [hunk 6]
```diff
@@ -10,30 +10,56 @@ accounting file = /var/log/tac_plus.acct
...
+        permit .*
+    }
+    cmd = /usr/bin/python {
+        deny .*
+    }
+    cmd = /usr/bin/find {
+        deny   -exec
+        permit .*
+    }
+    cmd = /usr/lib/x86_64-linux-gnu/ld-version.so {
```
Comment (qiluo-msft) [inline tests/tacacs/tac_plus.conf.j2:59]: >version [](http://example.com/codeflow?start=39&length=7)

I cannot find this file on a Debian 10. Is it expected?

File: tests/kvmtest.sh [hunk 7]
```diff
@@ -130,6 +130,7 @@ test_t0() {
     tacacs/test_ro_disk.py \
     tacacs/test_jit_user.py \
     tacacs/test_authorization.py \
+    tacacs/test_accounting.py \
```
Comment (qiluo-msft) [inline tests/kvmtest.sh:133]: Why changing the permission of this script?

File: tests/tacacs/tac_plus.conf.j2 (hunk 6 above)
Comment (qiluo-msft) [reply tests/tacacs/tac_plus.conf.j2]: I see you use this name as a placeholder. How about `/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2` which is a existing file and make more sense?

**Review by qiluo-msft — APPROVED** [review]

**Comment by liuh-80** [comment]
Because the accounting feature depends on the sonic-build-image side code, so need wait for the image used by 'checks' update.

**Comment by liuh-80** [comment]
Add per-command authorization code back, and use following PR to test the ignore older version code.

File: tests/tacacs/utils.py [hunk 8]
```diff
@@ -66,11 +112,23 @@ def setup_tacacs_server(ptfhost, creds_all_duts, duthost):
...
 
     # reset tacacs client configuration
-    duthost.shell("sudo config tacacs delete %s" % tacacs_server_ip)
+    remove_all_tacacs_server(duthost)
     duthost.shell("sudo config tacacs default passkey")
     duthost.shell("sudo config aaa authentication login default")
     duthost.shell("sudo config aaa authentication failthrough default")
+
+    (skip, reason) = check_skip_release(duthost, per_command_check_skip_versions)
+    if skip:
```
Comment (qiluo-msft) [inline tests/tacacs/utils.py:124]: >skip [](http://example.com/codeflow?start=7&length=4)

Wondering if you intent to `not skip` ?

File: tests/tacacs/utils.py [hunk 9]
```diff
@@ -66,11 +112,23 @@ def setup_tacacs_server(ptfhost, creds_all_duts, duthost):
...
+    stop_tacacs_server(ptfhost)
 
     # reset tacacs client configuration
-    duthost.shell("sudo config tacacs delete %s" % tacacs_server_ip)
+    remove_all_tacacs_server(duthost)
     duthost.shell("sudo config tacacs default passkey")
     duthost.shell("sudo config aaa authentication login default")
     duthost.shell("sudo config aaa authentication failthrough default")
+
+    (skip, reason) = check_skip_release(duthost, per_command_check_skip_versions)
```
Comment (qiluo-msft) [inline tests/tacacs/utils.py:123]: >(skip, reason) [](http://example.com/codeflow?start=4&length=14)

If you don't care about reason, you may use `skip, _` to simplify code.

**Review by qiluo-msft — DISMISSED** [review]

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**